data/src/main/java/com/microsoft/azure/kusto/data/Utils.java

```diff
@@ -154,22 +166,35 @@ private static DataServiceException createExceptionFromResponse(String url, Http
              *   result), or (2) in the KustoOperationResult's QueryCompletionInformation, both of which present with "200 OK". See .Net's DataReaderParser.
              */
             String activityId = determineActivityId(httpResponse);
+            String message;
+            WebException formattedException;
+            boolean isPermanent = false;
             if (!StringUtils.isBlank(errorFromResponse)) {
-                String message;
-                DataWebException formattedException = new DataWebException(errorFromResponse, httpResponse);
                 try {
-                    message = String.format("%s, ActivityId='%s'", formattedException.getApiError().getDescription(), activityId);
-                    return new DataServiceException(url, message, formattedException,
-                            formattedException.getApiError().isPermanent());
-                } catch (Exception ignored) {
+                    JSONObject jsonObject = new JSONObject(errorFromResponse);
+                    if (jsonObject.has("error")) {
+                        formattedException = new DataWebException(errorFromResponse, httpResponse, thrownException);
+                        message = String.format("%s, ActivityId='%s'", ((DataWebException)formattedException).getApiError().getDescription(), activityId);
+                        isPermanent = ((DataWebException)formattedException).getApiError().isPermanent();
+                    } else if (jsonObject.has("message")) {
+                        formattedException = new WebException(errorFromResponse, httpResponse, thrownException);
+                        message = String.format("%s, ActivityId='%s'", jsonObject.getString("message"), activityId);
+                    } else {
+                        formattedException = new WebException(errorFromResponse, httpResponse, thrownException);
+                        message = String.format("%s, ActivityId='%s'", errorFromResponse, activityId);
```

Comment:
Same as the catch, you can reduce cases

Reply:
I reduced a lot of the logic, let me know if that's ok.